osu.Game/Rulesets/Difficulty/PerformanceCalculator.cs

```diff
@@ -13,23 +14,25 @@ namespace osu.Game.Rulesets.Difficulty
 {
     public abstract class PerformanceCalculator
     {
-        protected readonly DifficultyAttributes Attributes;
+        protected readonly IEnumerable<TimedDifficultyAttributes> TimedAttributes;
```

Comment:
Any specific reason why this is IEnumerable instead of IList? 
With current implementation each Calculate(time) call results in TimedAttributes getting recalculated (CalculateTimed) with is quite slow.

Reply:
Why would it be an IList? IEnumerable is lazy, so you can just call ToList() or ToArray() on it anyway. Also, Calculate calculates for the end of the beatmap, not some arbitrary time. If you want to get values for specific points in the beatmap, you can enumerate the IEnumerable.

The only downside I could see is some more heap allocations for TimedDifficultyAttributes, but you can probably fix that by making it a valuetype/valuetuple and I doubt it'll be a problem in the first place.

Reply:
`just call ToList() or ToArray() on it anyway.` I can't since it's protected, and I wouldn't want to do that anyway as it's implementation detail of the class with users shouldn't have to deal with elsewhere.

`Calculate calculates for the end of the beatmap, not some arbitrary time.`
https://github.com/ppy/osu/pull/3909/files/d4f4f5f4f98011e66d7e9fdf0d333a64dfc2512d#diff-10f70d41626d6c8006cdb456da9bc44cR49
If I understand correctly it returns beatmap attributes at specific point of time. The way it is used in the function I linked makes it not lazy(whole IEnumerable is being evaluated each time) and thus slow and wastefull.

Reply:
Oh, my bad, I thought you were talking about DifficultyCalculator, not PerformanceCalculator.